source/common/upstream/upstream_impl.h

```diff
@@ -271,9 +272,18 @@ class HostDescriptionImpl : virtual public HostDescription,
                                 const envoy::config::core::v3::Metadata* metadata) const;
   absl::optional<MonotonicTime> lastHcPassTime() const override { return last_hc_pass_time_; }
 
-  void setAddressList(const std::vector<Network::Address::InstanceConstSharedPtr>& address_list) {
-    address_list_ = address_list;
-    ASSERT(address_list_.empty() || *address_list_.front() == *address_);
+  void setAddressList(const std::shared_ptr<std::vector<Network::Address::InstanceConstSharedPtr>>&
+                          address_list_ptr) {
+    address_list_ = address_list_ptr;
+    ASSERT(address_list_->empty() || *address_list_->front() == *address_);
+  }
+
+  const std::shared_ptr<std::vector<Network::Address::InstanceConstSharedPtr>>
+  copyAddressestoMem(const std::vector<Network::Address::InstanceConstSharedPtr>& address_list) {
```

Comment:
I might be missing something, but why is this needed here, and why is it called in many places?
It seems that this PR already updates many places that call the setter, and maybe only https://github.com/envoyproxy/envoy/blob/9c379377598f32b51d67d020e2ed5b98dc913ee4/source/common/upstream/upstream_impl.cc#L2067 needs to add a copy to avoid bubbling this shared_ptr-const outwards (looking at the method signature for `PriorityStateManager::registerHostForPriority()` I think that if the passed vector is already const, then it should've probably been shared_ptr<const> in the first place, but that can be deferred to a later PR).

We can discuss this offline.

Reply:
Thanks for the comments!

We can certainly move the make_shared() and address vector copy operation to inside setAddressList(). That way we can keep it in one place.  However setAddressList() called in LogicalHost->setNewAddresses() is under the write lock:
https://github.com/envoyproxy/envoy/blob/c9fb17a6fabd1d2030ce8f5192780e947a335e01/source/extensions/clusters/common/logical_host.h#L47

We need to minimize the operations under write lock as much as well. Thus we moved those operations out of the write lock, and have the setAddressList(), which is under write lock, just copy a shared_ptr.  This change actually improves the performance a little bit during DNS refreshing.

Reply:
For your 2nd part of the comments, changing the caller to pass in shared_ptr in the first place, that will trigger changes in a lot of places from DFP cluster to logical cluster. let's delay that kind of refactoring to later.